From 74654ec91e9e843cd28328557a938b5c3eef328b Mon Sep 17 00:00:00 2001
From: Marek Marczykowski <marmarek@invisiblethingslab.com>
Date: Fri, 5 Apr 2013 01:37:29 +0200
Subject: [PATCH] conf: add 'script' attribute to disk specification
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Script to be called to prepare custom device for domain. Done with Xen
in mind, it maps to libxl_device_disk.script.

XML configuration would be:
<disk type='block' device='disk'>
    <source dev='/dev/mapper/custom-device'/>
    <script path='/script/to/setup/custom-device'/>
    <target dev='xvdc'/>
</disk>

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
---
 src/conf/domain_conf.c            | 3 +++
 src/conf/domain_conf.h            | 1 +
 src/conf/schemas/domaincommon.rng | 8 ++++++++
 3 files changed, 12 insertions(+)

diff --git a/src/conf/domain_conf.c b/src/conf/domain_conf.c
index 94a1f8bd8e..776d72b727 100644
--- a/src/conf/domain_conf.c
+++ b/src/conf/domain_conf.c
@@ -2470,6 +2470,7 @@ virDomainDiskDefFree(virDomainDiskDef *def)
     g_free(def->domain_name);
     g_free(def->blkdeviotune.group_name);
     g_free(def->virtio);
+    g_free(def->script);
     virDomainDeviceInfoClear(&def->info);
     virObjectUnref(def->privateData);
     g_slist_free_full(def->iothreads, (GDestroyNotify) virDomainIothreadMappingDefFree);
@@ -8890,6 +8891,7 @@ virDomainDiskDefParseXML(virDomainXMLOption *xmlopt,
         return NULL;
 
     def->domain_name = virXPathString("string(./backenddomain/@name)", ctxt);
+    def->script = virXPathString("string(./script/@path)", ctxt);
     def->serial = virXPathString("string(./serial)", ctxt);
     def->wwn = virXPathString("string(./wwn)", ctxt);
     def->vendor = virXPathString("string(./vendor)", ctxt);
@@ -24853,6 +24855,7 @@ virDomainDiskDefFormat(virBuffer *buf,
         return -1;
 
     virBufferEscapeString(&childBuf, "<backenddomain name='%s'/>\n", def->domain_name);
+    virBufferEscapeString(&childBuf, "<script path='%s'/>\n", def->script);
 
     virDomainDiskGeometryDefFormat(&childBuf, def);
     virDomainDiskBlockIoDefFormat(&childBuf, def);
diff --git a/src/conf/domain_conf.h b/src/conf/domain_conf.h
index eadb5a49a4..ec36b278d9 100644
--- a/src/conf/domain_conf.h
+++ b/src/conf/domain_conf.h
@@ -612,6 +612,7 @@ struct _virDomainDiskDef {
     unsigned int queue_size;
     virDomainDiskModel model;
     virDomainVirtioOptions *virtio;
+    char *script;
     virTristateSwitch dpofua; /* only SCSI disks; control of the DPO/FUA cache setting */
 
     bool diskElementAuth;
diff --git a/src/conf/schemas/domaincommon.rng b/src/conf/schemas/domaincommon.rng
index d1b10c1c5f..6009b0a5e2 100644
--- a/src/conf/schemas/domaincommon.rng
+++ b/src/conf/schemas/domaincommon.rng
@@ -1708,6 +1708,14 @@
           <empty/>
         </element>
       </optional>
+      <optional>
+        <element name="script">
+          <attribute name="path">
+            <ref name="filePath"/>
+          </attribute>
+          <empty/>
+        </element>
+      </optional>
       <optional>
         <element name="readonly">
           <empty/>
-- 
2.54.0

